pkg/cmd/pr/merge/merge.go

```diff
@@ -196,16 +196,25 @@ func mergeRun(opts *MergeOptions) error {
 		return cmdutil.SilentError
 	}
 
+	if !opts.AutoMergeEnable && pr.MergeStateStatus == "BLOCKED" {
+		fmt.Fprintf(
+			opts.IO.ErrOut,
+			"%s Merging is blocked. Enable auto-merge to automatically merge it when all requirements are met with: gh pr merge %d --auto\n",
+			cs.FailureIcon(), pr.Number)
+		return cmdutil.SilentError
+	}
+
 	deleteBranch := opts.DeleteBranch
 	crossRepoPR := pr.HeadRepositoryOwner.Login != baseRepo.RepoOwner()
+	autoMerge := opts.AutoMergeEnable && pr.MergeStateStatus == "BLOCKED"
```

Comment:
I think this conditional is not restrictive enough based on the potential values of `MergeStateStatus`. According to https://docs.github.com/en/graphql/reference/enums#mergestatestatus I feel like we should only try to actually merge the PR when the  `MergeStateStatus` is one of `UNSTABLE `, `HAS_HOOKS`, or `CLEAN`. Having a whitelist approach also protects us from any updates in new values are added to `MergeStateStatus`.

Reply:
Sorry for the late reply, the only thing the `MergeStateStatus` field is used for is to determine if a PR is in the "BLOCKED" state (in which case it enables auto-merge).

Currently if you try to merge a PR that has some conflict (either a merge conflict or checks failing), gh warns you that there's a conflict:
https://github.com/cli/cli/blob/33c3fb5cdd110f08791bf6afbdad522c414e7a54/pkg/cmd/pr/merge/merge.go#L193-L198

Do you think there should be an additional check based only on `MergeStateStatus`? If so, what benefits would it provide over the current one relying on the `Mergeable` field?

Reply:
Sorry for my confusing comment above. I think I could have better illustrated it with code.

```golang
	approved := pr.MergeStateStatus == "CLEAN" || pr.MergeStateStatus == "HAS_HOOKS" || pr.MergeStateStatus == "UNSTABLE"
	if !opts.AutoMergeEnable && !approved {
		fmt.Fprintf(
			opts.IO.ErrOut,
			"%s Merging is blocked. Enable auto-merge to automatically merge it when all requirements are met with: gh pr merge %d --auto\n",
			cs.FailureIcon(), pr.Number)
		return cmdutil.SilentError
	}

	deleteBranch := opts.DeleteBranch
	crossRepoPR := pr.HeadRepositoryOwner.Login != baseRepo.RepoOwner()
	autoMerge := opts.AutoMergeEnable && !approved
```
The `Mergeable` field, as you stated, is already checked on a previous line but that only checks that `git` can actually merge the PR without conflicts. The `MergeStateStatus` checks that branch checks have also passed, both of these checks are necessary to merge a PR. Values of `CLEAN`, `HAS_HOOKS`, and `UNSTABLE` are the values of `MergeStateStatus` that mean the branch checks have passed. So rather than than say `MergeStateStatus` != `BLOCKED` I would rather check that `MergeStateStatus` is equal to one of the passing values.